contrib/babelfishpg_tsql/src/tsqlIface.cpp

```diff
@@ -5764,12 +5815,129 @@ makeGrantdbStatement(TSqlParser::Security_statementContext *ctx)
+				if (single_perm->INSERT())
+					privilege_list = lappend(privilege_list, (void *)"insert");
+				if (single_perm->UPDATE())
+					privilege_list = lappend(privilege_list, (void *)"update");
+				if (single_perm->DELETE())
+					privilege_list = lappend(privilege_list, (void *)"delete");
+				if (single_perm->REFERENCES())
+					privilege_list = lappend(privilege_list, (void *)"references");
+			}
+			result->privileges = privilege_list;
+			if (ctx->grant_statement()->WITH())
+				result->with_grant_option = true;
+			result->grantees = grantee_list;
+			return (PLtsql_stmt *) result;
+		}
+	}
+
+	if (ctx->revoke_statement() && ctx->revoke_statement()->ON() && ctx->revoke_statement()->permission_object()
+			&& ctx->revoke_statement()->permission_object()->object_type() && ctx->revoke_statement()->permission_object()->object_type()->SCHEMA())
+	{
+		if (ctx->revoke_statement()->FROM() && ctx->revoke_statement()->principals() && ctx->revoke_statement()->permissions())
+		{
+			PLtsql_stmt_grantschema *result = (PLtsql_stmt_grantschema *) palloc0(sizeof(PLtsql_stmt_grantschema));
+			result->cmd_type = PLTSQL_STMT_GRANTSCHEMA;
+			result->lineno = getLineNo(ctx->revoke_statement());
+			result->is_grant = false;
+			std::string schema_name;
+			if (ctx->revoke_statement()->permission_object()->full_object_name()->object_name)
+			{
+				schema_name = stripQuoteFromId(ctx->revoke_statement()->permission_object()->full_object_name()->object_name);
+				result->schema_name = pstrdup(downcase_truncate_identifier(schema_name.c_str(), schema_name.length(), true));
+			}
+			List *grantee_list = NIL;
+			for (auto prin : ctx->revoke_statement()->principals()->principal_id())
+			{
+				if (prin->id())
+				{
+					std::string id_str = ::getFullText(prin->id());
+					char *grantee_name = pstrdup(downcase_truncate_identifier(id_str.c_str(), id_str.length(), true));
+					grantee_list = lappend(grantee_list, grantee_name);
+				}
+				if (prin->PUBLIC())
+				{
+					char *grantee_name = pstrdup("public");
+					grantee_list = lappend(grantee_list, grantee_name);
+				}
+			}
+			List *privilege_list = NIL;
+			for (auto perm: ctx->revoke_statement()->permissions()->permission())
+			{
+				auto single_perm = perm->single_permission();
+				if (single_perm->EXECUTE())
+					privilege_list = lappend(privilege_list, (void *)"execute");
+				if (single_perm->EXEC())
+					privilege_list = lappend(privilege_list, (void *)"execute");
+				if (single_perm->SELECT())
+					privilege_list = lappend(privilege_list, (void *)"select");
+				if (single_perm->INSERT())
+					privilege_list = lappend(privilege_list, (void *)"insert");
+				if (single_perm->UPDATE())
+					privilege_list = lappend(privilege_list, (void *)"update");
+				if (single_perm->DELETE())
+					privilege_list = lappend(privilege_list, (void *)"delete");
+				if (single_perm->REFERENCES())
+					privilege_list = lappend(privilege_list, (void *)"references");
```

Comment:
Why do we need to store mask values in `PLtsql_stmt_grantschema` instead of char*? 
Based on the privileges in the list, we generate the GRANT/ALTER PRIVILEGE queries. There will be multiple queries  for different privilege since the object type where it is applicable can be different.

If we do bitmask here and store the int value in the catalog, again we would need to get the list of privileges to generate GRANT schema subcommands in `gen_grantschema_subcmds`. 

IMO this seems better. Please suggest if there is a better way.

Reply:
You are already doing string comparison for gen_grantschema_subcmds. Storing a simple integer make MUCH more SENSE. Dont forget the benefit you get later during storing. :) 
You are then gonna compare atomically which is SUPER CHEAP in Time and space complexity, aint it? :)

Reply:
Done.